cmd/gcp-controller-manager/node_syncer.go

```diff
@@ -156,13 +191,36 @@ func (ns *nodeSyncer) process(key string) error {
 	if err != nil {
 		return fmt.Errorf("failed to get Pod %q: %v", key, err)
 	}
-	if !exists {
-		node, gsa, found := ns.nodes.remove(key)
-		if !found {
+	if !exists { // pod removal event
+		podUIDs := ns.nodes.find(key)
+		if len(podUIDs) == 0 {
+			klog.Warningf("Pod key %q not found", key)
+			return nil // no retry
+		}
+		if ns.delayGSARemove {
+			for _, uid := range podUIDs {
+				klog.Infof("Pod %q (UID %q) is queued for delayed removal", key, uid)
+				ns.podRemoveQueue.AddAfter(uid, nodeSyncerGSARemoveDelay)
+			}
 			return nil
 		}
-		klog.Infof("Pod %q and its GSA %q is removed from Node %q", key, gsa, node)
-		return ns.sync(node)
+		nodeSet := make(map[string]bool)
+		for _, uid := range podUIDs {
+			node, gsa, found := ns.nodes.remove(uid)
+			if !found {
+				klog.Warning("Pod %q (UID %q) not found on removal event: %v", key, uid, err)
+				continue
+			}
+			klog.Infof("Pod %q running as %q is removed from Node %q (pod UID: %q)", key, gsa, node, uid)
+			nodeSet[node] = true
+		}
+		for node := range nodeSet {
+			if err := ns.sync(node); err != nil {
+				// Log only; retries will be triggered by informer's resync events.
+				klog.Warningf("Failed to sync Node %q for GSA removal: %v", node, err)
+			}
+		}
+		return nil
```

Comment:
Why is so much changing outside of  ns.delayGSARemove == true branches?

Reply:
It is because pod entries are now stored by their UIDs which is later used as the key to the delay queue.  UID is used instead of ns/name such that pods of the name will be delay-queued and handled individually (with the same or different target GSA).  As such, the original logics now also have to look up UID before inserting the pod (or in abnormal case pods) to the map.